### PR TITLE
Allow specialized engulfing in microbe colony

### LINKED ISSUE #3549: If primary cell in a colony can't engulf, but other members can they should be able to enter engulf mode
Right now the primary cell in a colony (at least when controlled by the player) must have a membrane allowing engulfing, otherwise no member in the colony can enter engulf mode. We should update things so that the player can enter engulf mode and is shown the hot bar icon if any cell in their colony can engulf.

### PR DESCRIPTION
**Pull request by Kasterisk**
**Brief Description of What This PR Does**

Walled cells can now be in engulf state but can't act on it, this is to allow specialized engulfing in a colony.

I had to make two ways of checking whether a cell is capable to engulf, one is through `ICellProperties` and the other through `Microbe` itself, decoupled from species. The reason for this is because `ICellProperties` sometimes may not reflect the actual properties visible in a cell.

I hadn't yet tested in multicellular but this should work fine for colony.

**Related Issues**

Fixes #3549 

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
Why did you not take the approach that when a cell tries to enter engulf mode, even if it cannot do it, it would broadcast the message to the colony? That way any cells that *can* engulf in the colony would switch to engulf mode. I thought that was the pretty obvious solution to #3549

**Comment by Kasterisk** [comment]
That's what I did.

https://github.com/Revolutionary-Games/Thrive/blob/ee6e5c7b5ee1bab6aee56607c955ff07f78ca53f/src/microbe_stage/Microbe.Contact.cs#L1122-L1126

**Comment by hhyyrylainen** [comment]
I saw quite a lot of changes when skimming the diff. I suppose I'll look at it properly.

File: src/microbe_stage/Microbe.Contact.cs [hunk 1]
```diff
@@ -491,6 +460,16 @@ public bool CanEngulf(IEngulfable target)
         return EngulfSize > target.EngulfSize * Constants.ENGULF_SIZE_RATIO_REQ;
     }
 
+    /// <summary>
+    ///   Just like <see cref="ICellProperties.CanEngulf"/> but decoupled from Species and is based on the local
+    ///   condition of the microbe instead.
+    /// </summary>
+    /// <returns>True if this cell has all the requirements it needs for engulfing.</returns>
+    public bool CheckEngulfCapability()
```
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.Contact.cs:468]: I think this should be a computed property, doesn't make sense to me to wrap this tiny bit of code in a method.

File: src/microbe_stage/MicrobeColony.cs [hunk 2]
```diff
@@ -8,10 +8,11 @@
 [UseThriveSerializer]
 public class MicrobeColony
 {
-    private Microbe.MicrobeState state;
+    private MicrobeState state;
 
-    private bool hexCountDirty = true;
+    private bool memberDirty = true;
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeColony.cs:13]:
```suggestion
    private bool membersDirty = true;
```

File: src/microbe_stage/Microbe.Contact.cs (hunk 1 above)
Comment (hhyyrylainen) [reply src/microbe_stage/Microbe.Contact.cs:468]: Also why isn't this named `CanEngulf` like in the species, that seems a bit confusing to me to name the same thing multiple different things?

File: src/microbe_stage/Microbe.Contact.cs [hunk 3]
```diff
@@ -235,6 +189,21 @@ public float EngulfSize
         }
     }
 
+    /// <summary>
+    ///   Returns true if this microbe or the colony this microbe is part of has the capability to engulf.
+    /// </summary>
+    [JsonIgnore]
+    public bool HasEngulfCapability
```
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.Contact.cs:196]: I think this *really* needs a different name, I immediately confused this with `CheckEngulfCapability` while reading the changes here. Maybe something like `MakesSenseToAllowEngulfMode` or? something like that that implies that this cell doesn't necessary have the ability to engulf but still it makes sense for it to try to do so.

Comment (Kasterisk) [reply src/microbe_stage/Microbe.Contact.cs]: Renamed.

File: src/microbe_stage/Microbe.Contact.cs (hunk 1 above)
Comment (Kasterisk) [reply src/microbe_stage/Microbe.Contact.cs]: Fixed.

File: src/microbe_stage/MicrobeColony.cs [hunk 4]
```diff
@@ -172,4 +187,17 @@ private void UpdateHexCount()
...
         }
     }
+
+    private void UpdateCanEngulf()
+    {
+        canEngulf = false;
+
+        foreach (var member in ColonyMembers)
+        {
+            canEngulf = member.CanEngulf;
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeColony.cs:197]: The compiler might optimize this, but I think
```c#
if (!member.CanEngulf)
    continue;
canEngulf = true;
break;
```
would be preferable to show that the canEngulf variable only really needs to be written to once.

File: src/microbe_stage/MicrobeAI.cs [hunk 5]
```diff
@@ -243,7 +243,7 @@ private void ChooseActions(Random random, MicrobeAICommonData data, Microbe? sig
         }
 
         // If there are no threats, look for a chunk to eat
-        if (!microbe.CellTypeProperties.MembraneType.CellWall)
+        if (microbe.CanEngulf)
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeAI.cs:246]: You don't need to solve this now, but could you put in a TODO comment here that it'd be nice for the AI to know about colonies here and still consider engulfing things if it is in a colony that can engulf something?

File: src/microbe_stage/Microbe.Contact.cs [hunk 6]
```diff
@@ -1142,6 +1115,9 @@ private void HandleUnbinding()
     /// </summary>
     private void HandleEngulfing(float delta)
     {
+        if (!CanEngulf)
```
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.Contact.cs:1118]: I think this early returning is too early as this skips clearing `attemptingToEngulf` and also it doesn't handle the gradual smooth turning of the engulf audio. I think the cleanest refactor would be to create a new variable `actuallyEngulfing = State == MicrobeState.Engulf && CanEngulf` and replacing the existing checks in this method with that variable.

Comment (Kasterisk) [reply src/microbe_stage/Microbe.Contact.cs:1118]: Good idea, I'll fix this in a bit.

File: src/microbe_stage/MicrobeAI.cs (hunk 5 above)
Comment (Kasterisk) [reply src/microbe_stage/MicrobeAI.cs:247]: Added.

**Review by hhyyrylainen — APPROVED** [review]
Code change is fine now. Let's try to wait and see if anyone playtests this.

**Comment by hhyyrylainen** [comment]
I playtested this now and everything seemed to work fine with the colonies and single cells couldn't engulf things when they shouldn't.

**Pull request merged**